kentik_api_library/kentik_api/api_connection/api_connector.py

```diff
@@ -38,14 +53,38 @@ def send(self, api_call: APICall, payload: Optional[Dict[str, Any]] = None) -> A
         else:
             raise ValueError(f"Improper API call method: {api_call.method}")
 
-        self._validate_response(response)
+        # TODO(dfurman): consider catching and re-raising Requests exceptions
+
+        self._logger.debug("Got API response: %s", response)
+        self._handle_error(response)
 
         return APICallResponse(response.status_code, response.text)
 
     def _get_api_query_url(self, api_method: str) -> str:
         return self._api_url + api_method
 
-    def _validate_response(self, response: requests.Response) -> None:
-        # http error handling can be implemented here eg with exceptions, Expected[return, error] or error codes
+    @staticmethod
+    def _handle_error(response: requests.Response) -> None:
         if response.status_code >= 400:
-            self._logger.error("code: %d, body: %s", response.status_code, response.text)
+            raise new_api_error(response.text, response.status_code)
+
+
+def new_api_error(message: str, http_status_code: int) -> APIError:
+    common_errors = {
+        http.HTTPStatus.BAD_REQUEST: BadRequestError,
+        http.HTTPStatus.UNAUTHORIZED: UnauthorizedError,
+        http.HTTPStatus.FORBIDDEN: ForbiddenError,
+        http.HTTPStatus.NOT_FOUND: NotFoundError,
+        http.HTTPStatus.METHOD_NOT_ALLOWED: MethodNotAllowedError,
+        http.HTTPStatus.REQUEST_ENTITY_TOO_LARGE: PayloadTooLargeError,
+        http.HTTPStatus.UNSUPPORTED_MEDIA_TYPE: UnsupportedMediaTypeError,
+        http.HTTPStatus.TOO_MANY_REQUESTS: TooManyRequestsError,
+        http.HTTPStatus.INTERNAL_SERVER_ERROR: InternalServerError,
+        http.HTTPStatus.SERVICE_UNAVAILABLE: ServiceUnavailableError,
+        http.HTTPStatus.GATEWAY_TIMEOUT: GatewayTimeoutError,
+    }
+    try:
+        error = common_errors[http.HTTPStatus(http_status_code)](message, http_status_code)
+    except KeyError:
+        return APIError(message, http_status_code)
```

Comment:
does APIError returned here mean "We dont know what happened ?

Reply:
I decided to create custom errors only for common HTTP errors, such as `404 not found`. APIError is returned, when a status code is less common, e.g `418 I'm a teapot`. I was inspired by following set of errors: https://github.com/sendgrid/python-http-client/blob/main/python_http_client/exceptions.py#L65

Reply:
Do I merge this? I do not see draft status, but the conversation does not look resolved here... yet.

Let me know how you would like to proceed @Fenthick , @mateuszmidor

Reply:
@TedatKentik This is WiP, `Draft` label is under the title of the PR.